### PR TITLE
Implement Split Diff-View

### PR DESCRIPTION
**Pull request by bkcsoft**
Opened up a new PR since the previous one ( #2288) was for the wrong branch.

Changed the name from Unified => Split and rebased on develop

~~This is still a **WIP** so don't merge it yet, it's just here to show that I'm still working on it...~~

Stuff that still needs fixing before considering merging:
- [x] move inline styles to .less-files
- [x] testing (see https://github.com/gogits/gogs/pull/2288#issuecomment-167347361 and https://github.com/gogits/gogs/pull/2288#issuecomment-167424391)
- [x] move a few hard-coded strings from .tmpl-files to i18n

### CODE REVIEW

**Comment by bkcsoft** [comment]
@andreynering could you pull the latest and see if the problem is still there?
technically it was still side-by-side, just that the line was to wide and made it scroll (almost) forever ^.^ anyhow, I've made it line-wrap instead since it makes more sense for split-views :)

2016-01-01-233550_1600x900_scrot

2016-01-01-233558_1600x900_scrot

**Comment by bkcsoft** [comment]
Now working as intended

@Unknwon thoughts?

**Comment by Unknwon** [comment]
I think the button should be called `Unified` instead of `inline`.

**Comment by andreynering** [comment]
@bkcsoft Did you commit this change? Because for me it is as before.

I think you should rebase (or squash?) your commits, since some commits from @Unknwon is showing in this pull requests.

But I'd say this is working as expected.

**Comment by bkcsoft** [comment]
@Unknwon yeah, I think that makes most sense on second thought

@andreynering I've commited the changes, which web-browser are you testing in? (I'm currently only testing on firefox)
about the leaking commits, I rebased on develop and that seems to have borken something -. - I'll see if I can make it cleaner

**Comment by andreynering** [comment]
> Which web-browser are you testing in?

Chrome

**Comment by bkcsoft** [comment]
@andreynering yeah, not getting that on firefox or chrome. sure you checked it out cleanly and built it correctly? (`make build` and all that...)

@Unknwon Renamed and cleaned up a bit :)

**Comment by andreynering** [comment]
@bkcsoft I tried in Firefox and it's OK. In Chrome it's still as before.

**Comment by bkcsoft** [comment]
@andreynering which version if chrome? Since I canct reproduce it :/

**Comment by andreynering** [comment]
47.0.2526.106

**Comment by bkcsoft** [comment]
I'll have a look later today :)

**Comment by bkcsoft** [comment]
@andreynering currently testing in "Google Chrome 47.0.2526.80" and it "Works On My Machine" ^.^
Can't really see how .80 and .106 breaks it, I'll try 48.x or 49.x and see if it's still working there :)

**Comment by Unknwon** [comment]
Thanks @bkcsoft !

Just get back from trip. Will start reviewing today. (I'm on +8 CST right now)

**Comment by bkcsoft** [comment]
@andreynering odd, I tried it in "Google Chrome 47.0.2526.80" and there it worked as expected, but in "Google Chrome 47.0.2526.106" I see what you're seeing...

Anyhow, the issue is with word-wrapping... I wrap on whitespace, but there's no whitespace to wrap on :/

**Comment by andreynering** [comment]
I am playing a little here, trying to highlight what changed in each line, like GitHub and GitLab do:







There are some few bugs, but I hope I can get this working 100%.

**Comment by bkcsoft** [comment]
@andreynering well, I can make it either break in the middle of words, OR break on whitespace... apparently not both -.-

about highlighting changes, that might be _really_ hard... for simple stuff it easy enough:
```
- foo()
+ bar()
```
but for a bit more complex stuff it is harder:
```
- foo()
- bar(1)
+ foo()
+ baz()
+ bar(2)
```
^ this one should highlight only the numbers (and perhaps `baz()`) :smile:

**Comment by bkcsoft** [comment]
@andreynering well, I can make it either break in the middle of words, OR break on whitespace... apparently not both -.-

unless one uses an undocumented css-value... namely `word-break: break-word`...
I'll commit and push the fix so you can try it :smile:

**Comment by andreynering** [comment]
I will try to get it right and maybe send a PR, but will wait this one be
merged first.

I am using a third party package that does the diff of the strings.

Em dom, 3 de jan de 2016 22:14, Kim Carlbäcker <notifications@github.com>
escreveu:

**Comment by bkcsoft** [comment]
@andreynering yeah, I looked at diff-packages when I started this but it seemed like they wanted rad .patch-files (which I didn't wanna push to the frontend...) so I gave up on that...

**Comment by bkcsoft** [comment]
Verified to work in the following browsers:
Chrome stable 47.0.2526.106
Firefox 43.0.1
Microsoft Edge 20.10240.16384.0 (that version-number xD
Internet Explorer 11.0.10240.16431
Opera 33.0.1990,115

**Comment by andreynering** [comment]
@bkcsoft It's OK now. :+1:

**Comment by Unknwon** [comment]
Oops! There are conflicts on bindata.go

**Comment by Unknwon** [comment]
PS: A rebase is recommended to keep commit history clean.

File: routers/repo/commit.go [hunk 1]
```diff
@@ -168,6 +168,7 @@ func Diff(ctx *middleware.Context) {
 		}
 	}
 
+	ctx.Data["Style"] = ctx.Query("style") == "split"
```
Comment (Unknwon) [inline routers/repo/commit.go]: Shouldn't this be `IsSplitStyle` since it's a `bool` value?

File: routers/repo/commit.go [hunk 2]
```diff
@@ -213,6 +214,7 @@ func CompareDiff(ctx *middleware.Context) {
 	}
 	commits = models.ValidateCommitsWithEmails(commits)
 
+	ctx.Data["Style"] = ctx.Query("style") == "split"
```
Comment (Unknwon) [inline routers/repo/commit.go]: Same as above.

File: templates/repo/diff_box_unified.tmpl [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+											<td class="lines-code">
+												<pre>{{$line.Content}}</pre>
+											</td>
+										</tr>
+									{{end}}
+								{{end}}
+							</tbody>
+						</table>
+					</div>
+					<script>
```
Comment (Unknwon) [inline templates/repo/diff_box_unified.tmpl]: Should this part be removed?

File: templates/repo/diff_box_split.tmpl [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+							var prev = $(this).prev();
+							if(prev.is('.del-code') && prev.children().eq(3).text().trim() === '') {
+								while(prev.prev().is('.del-code') && prev.prev().children().eq(3).text().trim() === '') {
+									prev = prev.prev();
+								}
+								prev.children().eq(3).html($(this).children().eq(3).html());
+								prev.children().eq(2).html($(this).children().eq(2).html());
+								prev.children().eq(3).addClass('add-code');
+								prev.children().eq(2).addClass('add-code');
+//								prev.children().eq(2).attr('style', 'background-color: #eaffea !important');
```
Comment (Unknwon) [inline templates/repo/diff_box_split.tmpl]: What's going on with commented out lines?

**Comment by Unknwon** [comment]
Thanks agin for the PR!

I checked out locally and works great AFA I can tell.

Some comments are made, also I see a lot duplicated content between `diff_box_split.tmpl` and `diff_box_unified.tmpl`, is it worth doing it or should we reduce the duplicated part?

File: routers/repo/commit.go (hunk 1 above)
Comment (bkcsoft) [inline routers/repo/commit.go]: Indeed it should be IsSplitStyle

File: templates/repo/diff_box_split.tmpl (hunk 4 above)
Comment (bkcsoft) [inline templates/repo/diff_box_split.tmpl]: Old code, I'll fix that

File: templates/repo/diff_box_unified.tmpl (hunk 3 above)
Comment (bkcsoft) [inline templates/repo/diff_box_unified.tmpl]: Yes. Most've missed that when I moved this part to it's own file

**Comment by bkcsoft** [comment]
@Unknwon I've missed to remove to old code, I'll remove that and squash some commits to make it cleaner. Also resolve the bindata.go-issue

About the duplicate code, I made that decision because it'll only do 1 bool-check (and not 1 per line). My reasoning for that is that it should be faster to render, I can do some benchmarks if you'd like to see how much slower it'd be to do it inline. (just need to find a big commit first :smile:)

**Comment by Unknwon** [comment]
> About the duplicate code, I made that decision because it'll only do 1 bool-check (and not 1 per line). My reasoning for that is that it should be faster to render, I can do some benchmarks if you'd like to see how much slower it'd be to do it inline. (just need to find a big commit first :smile:)

I don't mean to check bool per line, but reduce the check to a smaller range of content. As you can see, things are getting different inside `tbody` tag: https://github.com/gogits/gogs/pull/2296/files#diff-9d1f05e66c594694d453ec18a343782dR35

**Comment by bkcsoft** [comment]
doing it on the `tbody`-tag would make it check `IsSplitStyle` on each listed file. but if that's acceptable I'll do that :)

**Comment by bkcsoft** [comment]
merging them and having it check per-file (on `tbody`) didn't even make a dent in the render-time even for some of the larger commits I could find :smile:

**Comment by bkcsoft** [comment]
Oh, I completely forgot to mention that this PR also includes the fold-icon (`opticon-fold`)
Which is what this is doing: https://github.com/gogits/gogs/pull/2296/files#diff-f0b73ac65b702903d90a117166a9b06bR75 (down to row 88)

Like this: 
fold-icon

Not sure if I should leave it there or have it in a separate PR?

**Comment by Unknwon** [comment]
Never mind, I'll do that after merge.

And you need at least comment out fold icon, it's not working yet.

**Comment by bkcsoft** [comment]
Already pushed the template-merging :)

The icon is more of a visual thing for me. but I'll remove it and see if I can make it work in a separate PR :)

**Comment by Unknwon** [comment]
Oh, I see. I'll checkout again later today. :grinning:

**Comment by Unknwon** [comment]
LGTM.

**Comment by mastorm** [comment]
Is there a flag to enable split-view by default?

**Comment by Unknwon** [comment]
@mastorm currently, no..

**Pull request merged**